Support for doctrine cache 2

mrVrAlex opened the pull request:
Feature https://github.com/Roave/psr-container-doctrine/issues/71

Ocramius reviewed (changes requested):
Check CI failures:

 * `composer.lock` needs to be updated
 * need to check that lowest dependency ranges are tested
 * potentially worth simply dropping `psr/cache:^1` and `doctrine/persistence:^2.3` entirely

mrVrAlex commented:
> potentially worth simply dropping psr/cache:^1 and doctrine/persistence:^2.3 entirel

Yeah, I think about it too. But drop persistence 2.3 means drop also ORM 2.11.
I planed for release 3.6 drop only doctrine/cache:^1 (allow only 2.0) and in future release (3.7?) can drop doctrine/persistence:^2.3

What you think?

Ocramius commented:
IMO just bump aggressively 🤔

Ocramius commented on composer.json:
```diff
@@ -18,20 +18,22 @@
         "doctrine/coding-standard": "^8.1",
         "phpunit/phpunit": "^9.4",
         "psalm/plugin-phpunit": "^0.16.1",
+        "symfony/cache": "^4.4 || ^5.4 || ^6.0",
```
Do we need all of these? We should probably use `^6`

Ocramius commented on composer.json:
```diff
@@ -18,20 +18,22 @@
         "phpunit/phpunit": "^9.4",
         "psalm/plugin-phpunit": "^0.16.1",
+        "symfony/cache": "^4.4 || ^5.4 || ^6.0",
+        "symfony/yaml": "^3.4 || ^4.0 || ^5.0 || ^6.0",
```
Do we need all of these? We should probably use `^6`

Ocramius commented on src/CacheFactory.php:
```diff
@@ -42,44 +32,7 @@ protected function createWithConfig(ContainerInterface $container, string $confi
         }
 
-        $instance = null;
-
-        if (array_key_exists('instance', $config)) {
-            $instance = is_string($config['instance']) ? $container->get($config['instance']) : $config['instance'];
-        }
-
-        switch ($config['class']) {
-            case FilesystemCache::class:
-            case PhpFileCache::class:
-                $cache = new $config['class']($config['directory']);
-                break;
-
-            case PredisCache::class:
-                assert($instance !== null);
-                $cache = new PredisCache($instance);
-                break;
-
-            case ChainCache::class:
-                $providers = array_map(
-                    function ($provider) use ($container): CacheProvider {
-                        return $this->createWithConfig($container, $provider);
-                    },
-                    is_array($config['providers']) ? $config['providers'] : []
-                );
-                $cache     = new ChainCache($providers);
-                break;
-
-            default:
-                $cache = $container->has($config['class']) ? $container->get($config['class']) : new $config['class']();
-        }
-
-        if ($cache instanceof MemcachedCache) {
-            assert($instance !== null);
-            $cache->setMemcached($instance);
-        } elseif ($cache instanceof RedisCache) {
-            assert($instance !== null);
-            $cache->setRedis($instance);
-        }
```
Hmm, this removal will break a lot of stuff :thinking:

mrVrAlex replied:
If we want stay this code, then need return support doctrine/cache 1.x version to composer.json. But in 2.x anyway these such classes do not exists, so update to 2.x will break cache configuration for doctrine and need write own PSR compatibility factory for cache providers.

Ocramius commented on src/ConfigurationFactory.php:
```diff
@@ -193,10 +160,10 @@ protected function createWithConfig(ContainerInterface $container, string $confi
+            //'metadata_cache' => 'array',
+            //'query_cache' => 'array',
+            //'result_cache' => 'array',
+            //'hydration_cache' => 'array',
```
Leftover?

Ocramius commented on src/ConfigurationFactory.php:
```diff
@@ -52,57 +52,23 @@ protected function createWithConfig(ContainerInterface $container, string $confi
+                    $cache,
+                    \Closure::fromCallable([$configuration, str_replace('_', '', ucwords('set_' . $cacheKey, '_'))])
+                );
+            }
```
Don't think this should be refactored this way - makes it much harder to understand.

Ocramius commented on src/DriverFactory.php:
```diff
@@ -49,7 +49,7 @@ protected function createWithConfig(ContainerInterface $container, string $confi
             $config['class'] !== AttributeDriver::class
             && ! is_subclass_of($config['class'], AttributeDriver::class)
-            && is_subclass_of($config['class'], AnnotationDriver::class)
+            && $config['class'] === AnnotationDriver::class
```
`is_subclass_of()` should probably stay? We're checking if `$config['class']` operates like an `AnnotationDriver`, not if it **IS** the `AnnotationDriver` :thinking:

Ocramius reviewed (changes requested):
CI failures indicate some trouble with PHP 7.4. Should we drop it? 

As for the rest: lots of breaking changes :scream: 

Can we somehow isolate this to a smaller change set, without affecting caching and similar?

PowerKiKi commented:
@mrVrAlex are you still willing to work on this PR ? or should the community step up to finish it ?

mrVrAlex commented:
> @mrVrAlex are you still willing to work on this PR ? or should the community step up to finish it ?

Yes, I will rebase this and will drop support doctrine persistence  2.x & cache 1.x as renovate bot want it)

Commit:
Bump doctrine/cache:^2.0

1. Doctrine (orm, dbal etc.) going to use PSR-6|PSR-16, so doctrine/cache drop own cache implementations which previously could be instantiated in CacheFactory according to config in "doctrine.cache.{$cacheType}". Now only 'class' key supported to either get as service from container or className to be instantiated.
2. ConfigurationFactory lost usage of cache by default, see "doctrine.configuration.{$entityManagerName}" keys:
  - 'metadata_cache'
  - 'query_cache'
  - 'result_cache'
  - 'hydration_cache'
  because no 'array' implementation
3. Left 'array' and 'filesystem' caches in full-config.php as example how to migrate to symfony/cache
4. Remove unused $configuration argument from ConfigurationFactory::processCacheImplementation()

PowerKiKi committed:
Use mocks instead of creating new class from scratch

PowerKiKi commented:
I created https://github.com/opsway/psr-container-doctrine/pull/2 to simplify this PR by not-adding `StubCache` when we can easily mock it with PHPUnit.

Let me know if you'd rather have in another format.

Slamdunk commented:
@mrVrAlex https://github.com/Roave/psr-container-doctrine/pull/95 has been merged and released, can you rebase this PR on top of the new `3.10.x` branch please?

mrVrAlex commented:
> @mrVrAlex #95 has been merged and released, can you rebase this PR on top of the new `3.10.x` branch please?

Yes, will do.

Ocramius commented on test/CacheFactoryTest.php:
```diff
@@ -167,4 +82,84 @@ public function testCanRetrieveCacheItemPoolFromContainer(): void
+        $container = $this->createMock(ContainerInterface::class);
+        $container
+            ->method('has')
+            ->withConsecutive(['config'], [$className])
```
We should avoid `withConsecutive()`: gone in PHPUnit 10

Ocramius commented on composer.json:
```diff
@@ -35,6 +34,12 @@
+    "suggest": {
+        "doctrine/cache": "Library deprecated, nevertheless Doctrine\\Common\\Cache\\Cache implementations wrapped into the bundled PSR-6 adapter"
+    },
+    "conflict": {
+        "doctrine/cache": "<2.0"
+    },
```
I don't think we need to conflict with this.

e-vil-dev replied:
ok, will remove to allow clients implement factories and continue use v1

Ocramius commented on example/full-config.php:
```diff
@@ -193,6 +137,16 @@
+                $params = $config['doctrine']['cache']['filesystem'];
+
+                return new FilesystemAdapter(
+                    namespace: $params['namespace'],
+                    directory: $params['directory'],
```
Would be best to avoid named arguments here, for compatibility with any potential upstream change.

The usage without named arguments seems very clear.

Ocramius commented on psalm.xml.dist:
```diff
@@ -53,5 +41,17 @@
                 <referencedMethod name="Doctrine\DBAL\Connection::getEventManager"/>
             </errorLevel>
         </DeprecatedMethod>
+        <UndefinedClass>
+            <errorLevel type="suppress">
+                <referencedClass name="Doctrine\Common\Cache\Cache"/>
+                <referencedClass name="Doctrine\Common\Cache\CacheProvider"/>
+                <referencedClass name="Doctrine\Common\Cache\Psr6\CacheAdapter"/>
+            </errorLevel>
+        </UndefinedClass>
+        <UndefinedDocblockClass>
+            <errorLevel type="suppress">
+                <referencedClass name="Doctrine\Common\Cache\Cache"/>
+            </errorLevel>
+        </UndefinedDocblockClass>
```
These should go away before merging: the fact that we reference these classes, yet they are not explicit dependencies, is a problem.

e-vil-dev replied:
those class referencesses are in `instanceof`-operator to check package availability (before replace cache with bundled `CacheAdapter`) and special `::class` consts in `InvalidArgumentException::fromUnsupportedCache()`. So as for me its safe for optional dependency.

Bytheway, until [doctrine/orm](https://github.com/doctrine/orm/blob/642a20bcff440234afe50c4a2f91c786298d5947/composer.json#L28) will not bump `doctrine/dbal:^4.0` all project will have `doctrine/cache`, but whats the reason to declare it as that package dependency.

As less error-prone approach I could inline suppression:
```
@psalm-suppress UndefinedClass, UnusedPsalmSuppress
```

Ocramius replied:
If we say that we go with `doctrine/cache:^2`, then we should most certainly not have any undefined classes at all.

If the support for `doctrine/cache` requires referencing classes from `doctrine/cache`, then the dependency should be declared in `composer.json`.

This applies to `instanceof` and `::class` references alike, but I'll check the patch in more detail when I have more time :)

e-vil-dev replied:
no its not about `doctrine/cache:^2` anymore. We allow v1 - https://github.com/Roave/psr-container-doctrine/pull/72#discussion_r1098606375 - but drop logic specific for this version (i mean instantiation support in `CacheFactory`). But stays logic for both versions:
- wrap into PSR-6 interface adapter required by ORM
- set namespace from `$config['doctrine']['cache'][$configKey]['namespace']` in case when no conventional cache service id [`"doctrine.cache.{$configKey}"`](https://github.com/Roave/psr-container-doctrine/blob/9cfad0cdd159ac86cd1d5d05759fb0007210b05d/src/AbstractFactory.php#L105) (assume its fully build for usage with adapter).

With knowledge that `doctrine/cache:^2` is an intermediate solution for migrating to the PSR-6 cache interface. It would be great to make package optional dependency.

internalsystemerror replied:
`doctrine/cache` should be moved from `require` to `require-dev` then, to satisfy psalm, and not impose it on users.

e-vil-dev replied:
yes, this will solve problem with ugly psalm suppression. It looks like ORM cache types are stable (metadata, query, result, hydration) so no problem that contributors may not aware about disappearing `doctrine/cache` package.

But there is other blocker for optional dependency for newbies - [composer-require-checker](https://github.com/Roave/psr-container-doctrine/blob/607b0ae0f8012eb91a598989959be05f1304437c/.github/workflows/composer-require-checker.yml#L27) - which require add `composer-require-checker.json` with list optional class lists:
```
{
  "symbol-whitelist" : [
    "Doctrine\\Common\\Cache\\Cache",
    "Doctrine\\Common\\Cache\\CacheProvider",
    "Doctrine\\Common\\Cache\\Psr6\\CacheAdapter"
  ]
}
```

e-vil-dev replied:
looks like `composer-require-checker` struggle against complexity caused by optional dependencies, so I will return `doctrine/cache` to `required` dependencies

Ocramius replied:
s/confused/correct.

The idea of `composer-require-checker` is precisely to disallow the existence of symbols that aren't part of the dependency set.

Ocramius commented on src/CacheFactory.php:
```diff
@@ -42,126 +29,28 @@ protected function createWithConfig(ContainerInterface $container, string $confi
         }
 
-        $instance = null;
+        $cache = $container->has($config['class']) ? $container->get($config['class']) : new $config['class']();
```
Since we are no longer supporting the various cache instance types, `new $config['class']` only remains useful/relevant for something like a `NullCache`: we can probably always use `$container->get()` for all scenarios.

e-vil-dev replied:
also for old default cache implementation - 'array'. So its usefull to simplify migration by add config:
```
  'doctrine' => [
    'cache' => [
      'array' => ['class' => \Symfony\Component\Cache\Adapter\ArrayAdapter::class],
    ],
  ],
```
without forcing adding useless services into container

Ocramius replied:
This would be a sufficient approach, but a void cache is preferable.

While it may be tempting to use `array` as a default, it is risky to do so for caches other than DQL and metadata caches, because `array` can break in long-running processes.

For result caches, a void/null cache would be best as a default.

e-vil-dev replied:
I would let clients decide which caches to use for which purposes. Also caches are optional, so we can avoid create own implementation of void/null/blackhole cache or be locked to another vendor (e.g. symfony/cache).

Above in https://github.com/Roave/psr-container-doctrine/pull/72#discussion_r1098482535 its just example how simple could be migration for client who use our default config with 'array' cache and if we will left constructor call in `CacheFactory`. We have no dependency on `symfony/cache`.

Ocramius commented on src/ConfigurationFactory.php:
```diff
@@ -52,57 +52,61 @@ protected function createWithConfig(ContainerInterface $container, string $confi
+            );
+
+            $this->processCacheImplementation(
+                $queryCache,
+                $configuration->setQueryCache(...),
+            );
+        }
+
+        if (isset($config['result_cache'])) {
+            $resultCache = $this->retrieveDependency(
+                $container,
+                $config['result_cache'],
+                'cache',
+                CacheFactory::class
+            );
 
-        $this->processCacheImplementation(
-            $configuration,
-            $hydrationCache,
-            [$configuration, 'setHydrationCache'],
-        );
+            $this->processCacheImplementation(
+                $resultCache,
+                $configuration->setResultCache(...),
+            );
+        }
+
+        if (isset($config['hydration_cache'])) {
+            $hydrationCache = $this->retrieveDependency(
+                $container,
+                $config['hydration_cache'],
+                'cache',
+                CacheFactory::class
+            );
+
+            $this->processCacheImplementation(
+                $hydrationCache,
+                $configuration->setHydrationCache(...),
+            );
+        }
```
Would probably be best to **not** have these conditionals, and always fallback to a null/void cache for the basic setup?

e-vil-dev replied:
I believe that setter-injection is for optional dependencies, so I think no need to force clients to provide a default cache implementation

Ocramius replied:
Yes, but we can provide a default array or void cache for most of these, as stated below in https://github.com/Roave/psr-container-doctrine/pull/72#discussion_r1097132938

e-vil-dev replied:
our dependencies now have no guaranteed cache implementation, so we can't

internalsystemerror replied:
🤔 It's a shame `psr/cache` v2 doesn't come with a null/void cache implementation since that's no doubt going to be the same for all implementations. Long shot but could it be added there?

e-vil-dev replied:
ok, I will try add null cache implementation to activate optional cache and to simplify `ConfigurationFactory`. I do not think that `psr/cache` may get any implementation even if it useful in real world.

Ocramius commented on src/ConfigurationFactory.php:
```diff
@@ -193,10 +198,10 @@ protected function createWithConfig(ContainerInterface $container, string $confi
     protected function getDefaultConfig(string $configKey): array
     {
         return [
-            'metadata_cache' => 'array',
-            'query_cache' => 'array',
-            'result_cache' => 'array',
-            'hydration_cache' => 'array',
+            'metadata_cache' => null,
+            'query_cache' => null,
+            'result_cache' => null,
+            'hydration_cache' => null,
```
We should probably declare a null/void/blackhole cache default

Commit:
Add default NullCache for ORM Configuration config to simplify its factory

1. doctrine/orm Configuration class and its parent from doctrine/dbal does not support null in cache setters
2. To avoid `if` in ConfigurationFactory create NullCache implementation of PSR-6 interface.
Avoid Doctrine\Common\Cache\Psr6\CacheItem (create own NullCacheItem) because someday we can remove "doctrine/cache" dependency
3. No type-hinting in NullCache and item to be compatible with "psr/cache:^1.0.1"

mrVrAlex commented:
@Ocramius Any further remarks after last fixes?  Soon in near future need add support for DBAL v4.0 ))

smsalert-mobi commented:
Hello, what is the status of this PR ? when it will be released ?

Slamdunk commented:
Ping :100:

Slamdunk commented:
Merged in https://github.com/Roave/psr-container-doctrine/pull/110, thank you @mrVrAlex @e-vil-dev
